### PR TITLE
Add RIGHT Spark function

### PR DESCRIPTION
**Pull request by konjac**
Spark official spec: https://spark.apache.org/docs/2.3.0/api/sql/#right

Also made some refinement to docs/comments of LEFT.

### CODE REVIEW

**Review by mbasmanova — COMMENTED** [review]
@konjac Kun, welcome. Thank you for the contribution.

@rui-mo @PHILO-HE Would you help review this PR?

File: velox/docs/functions/spark/string.rst [hunk 1]
```diff
@@ -158,6 +158,11 @@ Unless specified otherwise, all functions return NULL if at least one of the arg
         SELECT replace('ABCabc', 'abc', ''); -- ABC
         SELECT replace('ABCabc', 'abc', 'DEF'); -- ABCDEF
 
+.. spark:function:: right(string, length) -> string
+
+    Returns the rightmost length characters from the ``string``.
```
Comment (rui-mo) [inline velox/docs/functions/spark/string.rst:163]: length -> \``length\``

File: velox/docs/functions/spark/string.rst [hunk 2]
```diff
@@ -158,6 +158,11 @@ Unless specified otherwise, all functions return NULL if at least one of the arg
...
+    If ``length`` is less or equal than 0 the result is an empty string.
```
Comment (rui-mo) [inline velox/docs/functions/spark/string.rst:164]: Could we provide some examples in the documentation?

Comment (rui-mo) [inline velox/docs/functions/spark/string.rst:164]: What is the behavior of length being larger that the size of string, could we clarify?

File: velox/functions/sparksql/String.h [hunk 3]
```diff
@@ -768,6 +768,61 @@ struct LeftFunction {
   }
 };
 
+/// right function
+/// right(string, length) -> string
+/// Returns the rightmost length characters from the string
+/// Return an empty string if length is less or equal than 0
```
Comment (rui-mo) [inline velox/functions/sparksql/String.h:774]: nit: comments end with dot.

File: velox/functions/sparksql/tests/StringTest.cpp [hunk 4]
```diff
@@ -266,6 +266,22 @@ TEST_F(StringTest, left) {
...
+TEST_F(StringTest, right) {
+  const auto right = [&](const std::optional<std::string>& str,
+                        const std::optional<int32_t>& length) {
+    return evaluateOnce<std::string>("right(c0, c1)", str, length);
+  };
+  EXPECT_EQ(right("example", -2), "");
+  EXPECT_EQ(right("example", 0), "");
+  EXPECT_EQ(right("example", 2), "le");
+  EXPECT_EQ(right("example", 7), "example");
+  EXPECT_EQ(right("example", 20), "example");
```
Comment (rui-mo) [inline velox/functions/sparksql/tests/StringTest.cpp:278]: Could we add test for empty input string like `right("", 20)`?

**Review by rui-mo — COMMENTED** [review]
Thanks!

File: velox/docs/functions/spark/string.rst (hunk 2 above)
Comment (konjac) [reply velox/docs/functions/spark/string.rst:170]: Added. Thanks~

Comment (konjac) [reply velox/docs/functions/spark/string.rst:170]: @rui-mo Update the doc to indicate the original string will be returned.

File: velox/functions/sparksql/tests/StringTest.cpp (hunk 4 above)
Comment (konjac) [reply velox/functions/sparksql/tests/StringTest.cpp:284]: Empty string test is returned.

**Review by PHILO-HE — COMMENTED** [review]
Looks good to me! Please revise the PR title to meet community's requirement:
https://github.com/facebookincubator/velox/blob/main/CONTRIBUTING.md#adding-scalar-functions

File: velox/docs/spark_functions.rst [hunk 5]
```diff
@@ -60,32 +60,33 @@ for :doc:`all <functions/spark/coverage>` functions.
...
-    :spark:func:`exp`                 :spark:func:`notequalto`          :spark:func:`xxhash64`                                                                                      
-    :spark:func:`filter`              :spark:func:`pmod`                :spark:func:`year`                                                                                          
+    :spark:func:`abs`                 :spark:func:`get_json_object`     :spark:func:`regexp_extract`          :spark:func:`bit_xor`                 :spark:func:`nth_value`
+    :spark:func:`acos`                :spark:func:`greaterthan`         :spark:func:`remainder`               :spark:func:`first` 
+    :spark:func:`acosh`               :spark:func:`greaterthanorequal`  :spark:func:`replace`                 :spark:func:`first_ignore_null` 
+    :spark:func:`add`                 :spark:func:`greatest`            :spark:func:`right`                   :spark:func:`last` 
+    :spark:func:`aggregate`           :spark:func:`hash`                :spark:func:`rlike`                   :spark:func:`last_ignore_null`
```
Comment (PHILO-HE) [inline velox/docs/spark_functions.rst:67]: The community prefers a bulk update for this file (I guess, if not, it can easily cause code conflict). Let's remove this. Thanks!

Comment (konjac) [reply velox/docs/spark_functions.rst]: @PHILO-HE Thank you for the suggestion! The file has been reverted.

File: velox/docs/functions/spark/coverage.rst [hunk 6]
```diff
@@ -82,7 +82,7 @@
...
     :spark:func:`abs`                          count_if                                   inline                                     nvl                                        sqrt                                           any                                            cume_dist                                
-    :spark:func:`acos                          count_min_sketch                           inline_outer                               nvl2                                       stack                                          approx_count_distinct                          dense_rank                               
+    :spark:func:`acos`                         count_min_sketch                           inline_outer                               nvl2                                       stack                                          approx_count_distinct                          dense_rank
```
Comment (rui-mo) [inline velox/docs/functions/spark/coverage.rst:85]: Is this a fix for the documentation? Maybe we can do that is a separate PR.

File: velox/docs/functions/spark/string.rst [hunk 7]
```diff
@@ -158,6 +164,17 @@ Unless specified otherwise, all functions return NULL if at least one of the arg
         SELECT replace('ABCabc', 'abc', ''); -- ABC
         SELECT replace('ABCabc', 'abc', 'DEF'); -- ABCDEF
 
+.. spark:function:: right(string, length) -> string
+
+    Returns the rightmost ``length`` characters from the ``string``.
+    If ``length`` is less or equal than 0 the result is an empty string.
+    If ``length`` is larger than the length of ``string``, return the original string.
```
Comment (rui-mo) [inline velox/docs/functions/spark/string.rst:171]: return -> returns

File: velox/functions/sparksql/String.h [hunk 8]
```diff
@@ -715,8 +715,9 @@ struct OverlayVarbinaryFunction : public OverlayFunctionBase {
 
 /// left function
 /// left(string, length) -> string
-/// Returns the leftmost length characters from the string
-/// Return an empty string if length is less or equal than 0
+/// Returns the leftmost length characters from the string.
+/// Return an empty string if length is less or equal than 0.
+/// Return the original string if length is larger than the string length.
```
Comment (rui-mo) [inline velox/functions/sparksql/String.h:720]: Return -> Returns

File: velox/functions/sparksql/String.h [hunk 9]
```diff
@@ -768,6 +769,62 @@ struct LeftFunction {
   }
 };
 
+/// right function
+/// right(string, length) -> string
+/// Returns the rightmost length characters from the string.
+/// Return an empty string if length is less or equal than 0.
+/// Return the original string if length is larger than the string length.
```
Comment (rui-mo) [inline velox/functions/sparksql/String.h:776]: Ditto.

File: velox/functions/sparksql/String.h [hunk 10]
```diff
@@ -768,6 +769,62 @@ struct LeftFunction {
...
+    if (length > numCharacters) {
+      length = numCharacters;
+    }
+
+    int32_t start = numCharacters - length + 1;
+
+    auto byteRange =
+        stringCore::getByteRange<isAscii>(input.data(), start, length);
+
+    // Generating output string
```
Comment (rui-mo) [inline velox/functions/sparksql/String.h:822]: nit: add `.` at the end of comments.

File: velox/functions/sparksql/tests/StringTest.cpp [hunk 11]
```diff
@@ -266,6 +269,25 @@ TEST_F(StringTest, left) {
   EXPECT_EQ(left("da\u6570\u636Eta", 30), "da\u6570\u636Eta");
 }
 
+TEST_F(StringTest, right) {
```
Comment (rui-mo) [inline velox/functions/sparksql/tests/StringTest.cpp:272]: Please arrange the tests following alphabetical order.

Comment (konjac) [reply velox/functions/sparksql/tests/StringTest.cpp]: Updated

File: velox/functions/sparksql/String.h (hunk 10 above)
Comment (konjac) [reply velox/functions/sparksql/String.h]: Updated in new iteration.

File: velox/functions/sparksql/String.h (hunk 9 above)
Comment (konjac) [reply velox/functions/sparksql/String.h]: Updated in new iteration.

File: velox/functions/sparksql/String.h (hunk 8 above)
Comment (konjac) [reply velox/functions/sparksql/String.h]: Updated in new iteration.

File: velox/docs/functions/spark/string.rst (hunk 7 above)
Comment (konjac) [reply velox/docs/functions/spark/string.rst]: Updated in new iteration.

File: velox/docs/functions/spark/coverage.rst (hunk 6 above)
Comment (konjac) [reply velox/docs/functions/spark/coverage.rst]: Reverted this line in new iteration.

File: velox/docs/functions/spark/string.rst [hunk 12]
```diff
@@ -82,8 +82,14 @@ Unless specified otherwise, all functions return NULL if at least one of the arg
 
 .. spark:function:: left(string, length) -> string
 
-    Returns the leftmost length characters from the ``string``.
+    Returns the leftmost ``length`` characters from the ``string``.
     If ``length`` is less or equal than 0 the result is an empty string.
+    If ``length`` is larger than the length of ``string``, return the original string.
```
[the remark this answers is not in the document]
Comment (konjac) [reply velox/docs/functions/spark/string.rst]: Updated in new iteration.

File: velox/functions/sparksql/String.h [hunk 13]
```diff
@@ -762,7 +763,63 @@ struct LeftFunction {
...
+  }
+};
+
+/// right function
+/// right(string, length) -> string
+/// Returns the rightmost length characters from the string.
+/// Returns an empty string if length is less or equal than 0.
+/// Returns the original string if length is larger than the string length.
+template <typename T>
+struct RightFunction {
```
Comment (rui-mo) [inline velox/functions/sparksql/String.h:778]: I just noticed Spark maintained a replacement for the Right function. In this case, probably it is converted as substring which is already supported. Could you help confirm? Thanks.
https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala#L2008-L2016

File: velox/functions/sparksql/tests/StringTest.cpp [hunk 14]
```diff
@@ -478,6 +481,25 @@ TEST_F(StringTest, replace) {
       "123\u6570data");
 }
 
+TEST_F(StringTest, right) {
+  const auto right = [&](const std::optional<std::string>& str,
+                        const std::optional<int32_t>& length) {
```
Comment (rui-mo) [inline velox/functions/sparksql/tests/StringTest.cpp:486]: Why do we make length as optional in the test? It seems it is a required argument for right function.

**Review by rui-mo — COMMENTED** [review]
Thanks.
